File: notepad_core/ios/Classes/NotepadCorePlugin.m
#import <CoreBluetooth/CoreBluetooth.h>
#import "NotepadCorePlugin.h"

const int GATT_HEADER_LENGTH = 3;

NSString *GSS_SUFFIX = @"0000-1000-8000-00805f9b34fb";

# pragma CBUUID+Extensions

@interface CBUUID (Extensions)
@end

@implementation CBUUID (Extensions)
- (NSString *)uuidStr {
    return self.UUIDString.lowercaseString;
}
@end

# pragma CBPeripheral+Extensions

@interface CBPeripheral (Extensions)
@end

@implementation CBPeripheral (Extensions)
- (CBCharacteristic *)getCharacteristic:(NSString *)characteristic ofService:(NSString *)service {
    NSUInteger serviceIndex = [self.services indexOfObjectPassingTest:^BOOL(CBService *obj, NSUInteger idx, BOOL *stop) {
        NSString *withGss = [NSString stringWithFormat:@"0000%@-%@", obj.UUID.uuidStr, GSS_SUFFIX];
        return [obj.UUID.uuidStr isEqualToString:service] || [withGss isEqualToString:service];
    }];
    NSArray<CBCharacteristic *> *characteristics = self.services[serviceIndex].characteristics;
    NSUInteger characteristicIndex = [characteristics indexOfObjectPassingTest:^BOOL(CBCharacteristic *obj, NSUInteger idx, BOOL *stop) {
        NSString *withGss = [NSString stringWithFormat:@"0000%@-%@", obj.UUID.uuidStr, GSS_SUFFIX];
        return [obj.UUID.uuidStr isEqualToString:characteristic] || [withGss isEqualToString:characteristic];
    }];
    return characteristics[characteristicIndex];
}

- (void)setNotifiable:(NSString *)bleInputProperty forCharacteristic:(NSString *)characteristic ofService:(NSString *)service {
    [self setNotifyValue:![bleInputProperty isEqualToString:@"disabled"]
       forCharacteristic:[self getCharacteristic:characteristic ofService:service]];
}
@end

# pragma NotepadCorePlugin

@interface NotepadCorePlugin () <CBCentralManagerDelegate, FlutterStreamHandler, CBPeripheralDelegate>
@property(nonatomic, strong) CBCentralManager *manager;
@property(nonatomic, strong) NSMutableDictionary<NSString *, CBPeripheral *> *discoveredPeripherals;
@property(nonatomic, strong) CBPeripheral *peripheral;

@property(nonatomic, strong) FlutterBasicMessageChannel *connectorMessage;
@property(nonatomic, strong) FlutterBasicMessageChannel *clientMessage;
@property(nonatomic, strong) FlutterEventSink scanResultSink;

@end

@implementation NotepadCorePlugin
+ (void)registerWithRegistrar:(NSObject <FlutterPluginRegistrar> *)registrar {
    [[NotepadCorePlugin alloc] initWithRegistrar:registrar];
}

- (instancetype)initWithRegistrar:(NSObject <FlutterPluginRegistrar> *)registrar {
    if (self = [super init]) {
        _manager = [[CBCentralManager alloc] initWithDelegate:self queue:nil];
        _discoveredPeripherals = [[NSMutableDictionary alloc] init];

        FlutterMethodChannel *methodChannel = [FlutterMethodChannel methodChannelWithName:@"notepad_core/method" binaryMessenger:[registrar messenger]];
        [registrar addMethodCallDelegate:self channel:methodChannel];
        [[FlutterEventChannel eventChannelWithName:@"notepad_core/event.scanResult" binaryMessenger:[registrar messenger]] setStreamHandler:self];
        _connectorMessage = [FlutterBasicMessageChannel messageChannelWithName:@"notepad_core/message.connector" binaryMessenger:[registrar messenger]];
        _clientMessage = [FlutterBasicMessageChannel messageChannelWithName:@"notepad_core/message.client" binaryMessenger:[registrar messenger]];
    }
    return self;
}

# pragma FlutterPlugin

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSLog(@"handleMethodCall %@", call.method);
    if ([call.method isEqualToString:@"isBluetoothAvailable"]) {
        BOOL b = _manager.state == CBManagerStatePoweredOn;
        result(@(b));
    } else if ([call.method isEqualToString:@"startScan"]) {
        [_manager scanForPeripheralsWithServices:nil options:nil];
        result(nil);
    } else if ([call.method isEqualToString:@"stopScan"]) {
        [_manager stopScan];
        result(nil);
    } else if ([call.method isEqualToString:@"connect"]) {
        NSString *deviceId = call.arguments[@"deviceId"];
        _peripheral = _discoveredPeripherals[deviceId];
        _peripheral.delegate = self;
        [_manager connectPeripheral:_peripheral options:nil];
        result(nil);
    } else if ([call.method isEqualToString:@"disconnect"]) {
        if (_peripheral)
            [_manager cancelPeripheralConnection:_peripheral];
        [self clean];
        result(nil);
        [_connectorMessage sendMessage:@{@"ConnectionState": @"disconnected"}];
    } else if ([call.method isEqualToString:@"discoverServices"]) {
        [_peripheral discoverServices:nil];
        result(nil);
    } else if ([call.method isEqualToString:@"setNotifiable"]) {
        NSString *service = call.arguments[@"service"];
        NSString *characteristic = call.arguments[@"characteristic"];
        NSString *bleInputProperty = call.arguments[@"bleInputProperty"];
        [_peripheral setNotifiable:bleInputProperty forCharacteristic:characteristic ofService:service];
        result(nil);
    } else if ([call.method isEqualToString:@"requestMtu"]) {
        NSUInteger mtu = [_peripheral maximumWriteValueLengthForType:CBCharacteristicWriteWithoutResponse];
        result(nil);
        NSLog(@"peripheral.maximumWriteValueLengthForType:CBCharacteristicWriteWithoutResponse %lu", (unsigned long) mtu);
        [_clientMessage sendMessage:@{@"mtuConfig": @(mtu + GATT_HEADER_LENGTH)}];
    } else if ([call.method isEqualToString:@"requestConnectionPriority"]) {
        // Ignore API for Android
        result(nil);
    } else if ([call.method isEqualToString:@"readValue"]) {
        NSString *service = call.arguments[@"service"];
        NSString *characteristic = call.arguments[@"characteristic"];
        [_peripheral readValueForCharacteristic:[_peripheral getCharacteristic:characteristic ofService:service]];
        result(nil);
    } else if ([call.method isEqualToString:@"writeValue"]) {
        NSString *service = call.arguments[@"service"];
        NSString *characteristic = call.arguments[@"characteristic"];
        FlutterStandardTypedData *value = call.arguments[@"value"];
        NSString *bleOutputProperty = call.arguments[@"bleOutputProperty"];
        enum CBCharacteristicWriteType type = [bleOutputProperty isEqualToString:@"withoutResponse"] ? CBCharacteristicWriteWithoutResponse : CBCharacteristicWriteWithResponse;
        [_peripheral writeValue:[value data]
              forCharacteristic:[_peripheral getCharacteristic:characteristic ofService:service]
                           type:type];
        result(nil);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

# pragma CBCentralManagerDelegate

- (void)centralManagerDidUpdateState:(CBCentralManager *)central {
    NSLog(@"centralManagerDidUpdateState %ld", (long) central.state);
    switch (central.state) {
        case CBManagerStatePoweredOff:
            [_connectorMessage sendMessage:@{@"BluetoothState": @"unavailable"}];
            break;
        case CBManagerStatePoweredOn:
            [_connectorMessage sendMessage:@{@"BluetoothState": @"available"}];
            break;
    }

}

- (void)centralManager:(CBCentralManager *)central didDiscoverPeripheral:(CBPeripheral *)peripheral advertisementData:(NSDictionary<NSString *, id> *)advertisementData RSSI:(NSNumber *)RSSI {
    NSLog(@"centralManager:didDiscoverPeripheral %@ %@", peripheral.name, peripheral.identifier);
    [_discoveredPeripherals setValue:peripheral forKey:peripheral.identifier.UUIDString];

    NSData *manufacturerData = advertisementData[CBAdvertisementDataManufacturerDataKey];
    if (_scanResultSink)
        _scanResultSink(@{
                @"name": advertisementData[@"kCBAdvDataLocalName"] != nil ? advertisementData[@"kCBAdvDataLocalName"] : peripheral.name ? peripheral.name : @"",
                @"deviceId": peripheral.identifier.UUIDString,
                @"manufacturerData": [FlutterStandardTypedData typedDataWithBytes:(manufacturerData ? manufacturerData : [NSData new])],
                @"rssi": RSSI,
        });
}

- (void)centralManager:(CBCentralManager *)central didConnectPeripheral:(CBPeripheral *)peripheral {
    if (peripheral != _peripheral) {
        NSLog(@"Probably MEMORY LEAK!");
        return;
    }
    NSLog(@"centralManager:didConnect %@", peripheral.identifier);
    [_connectorMessage sendMessage:@{@"ConnectionState": @"connected"}];
}

- (void)centralManager:(CBCentralManager *)central didDisconnectPeripheral:(CBPeripheral *)peripheral error:(nullable NSError *)error {
    if (peripheral != _peripheral) {
        NSLog(@"Probably MEMORY LEAK!");
        return;
    }
    NSLog(@"centralManager:didDisconnectPeripheral: %@ error: %@", peripheral.identifier, error);
    [self clean];
    [_connectorMessage sendMessage:@{@"ConnectionState": @"disconnected"}];
}

- (void)clean {
    _peripheral = nil;
}

# pragma FlutterStreamHandler

- (FlutterError *_Nullable)onListenWithArguments:(id _Nullable)arguments eventSink:(FlutterEventSink)events {
    NSString *name = [arguments objectForKey:@"name"];
    NSLog(@"NotepadCorePlugin onListenWithArguments：%@", name);
    if ([name isEqualToString:@"scanResult"]) {
        _scanResultSink = events;
    }
    return nil;
}

- (FlutterError *_Nullable)onCancelWithArguments:(id _Nullable)arguments {
    NSString *name = [arguments objectForKey:@"name"];
    NSLog(@"NotepadCorePlugin onCancelWithArguments：%@", name);
    if ([name isEqualToString:@"scanResult"]) {
        _scanResultSink = nil;
    }
    return nil;
}

# pragma CBPeripheralDelegate

- (void)peripheral:(CBPeripheral *)peripheral didDiscoverServices:(nullable NSError *)error {
    if (peripheral != _peripheral) {
        NSLog(@"Probably MEMORY LEAK!");
        return;
    }
    NSLog(@"peripheral: %@ didDiscoverServices: %@", peripheral.identifier, error);
    for (CBService *service in peripheral.services) {
        [peripheral discoverCharacteristics:nil forService:service];
    }
}

- (void)peripheral:(CBPeripheral *)peripheral didDiscoverCharacteristicsForService:(CBService *)service error:(nullable NSError *)error {
    if (peripheral != _peripheral) {
        NSLog(@"Probably MEMORY LEAK!");
        return;
    }
    for (CBCharacteristic *characteristic in service.characteristics) {
        NSLog(@"peripheral:didDiscoverCharacteristicsForService (%@, %@)", service.UUID.uuidStr, characteristic.UUID.uuidStr);
    }
    [self->_connectorMessage sendMessage:@{
            @"ServiceState": @"discovered",
            @"services": @[service.UUID.uuidStr],
    }];
}

- (void)peripheral:(CBPeripheral *)peripheral didUpdateNotificationStateForCharacteristic:(CBCharacteristic *)characteristic error:(nullable NSError *)error {
    if (peripheral != _peripheral) {
        NSLog(@"Probably MEMORY LEAK!");
        return;
    }
    NSLog(@"peripheral:didUpdateNotificationStateFor %@ %d", characteristic.UUID, characteristic.isNotifying);
    [_clientMessage sendMessage:@{@"characteristicConfig": characteristic.UUID.uuidStr}];
}

- (void)peripheral:(CBPeripheral *)peripheral didWriteValueForCharacteristic:(CBCharacteristic *)characteristic error:(nullable NSError *)error {
    if (peripheral != _peripheral) {
        NSLog(@"Probably MEMORY LEAK!");
        return;
    }
    NSLog(@"peripheral:didWriteValueForCharacteristic %@ %@ error: %@", characteristic.UUID.uuidStr, characteristic.value, error);
}

- (void)peripheral:(CBPeripheral *)peripheral didUpdateValueForCharacteristic:(CBCharacteristic *)characteristic error:(nullable NSError *)error {
    if (peripheral != _peripheral) {
        NSLog(@"Probably MEMORY LEAK!");
        return;
    }
    NSLog(@"peripheral:didUpdateValueForCharacteristic %@ %@ error: %@", characteristic.UUID.uuidStr, characteristic.value, error);
    NSDictionary *characteristicValue = @{
            @"characteristic": characteristic.UUID.uuidStr,
            @"value": [FlutterStandardTypedData typedDataWithBytes:characteristic.value],
    };
    [_clientMessage sendMessage:@{@"characteristicValue": characteristicValue}];
}

@end

File: notepad_core_macos/macos/Classes/NotepadCorePlugin.swift
import Cocoa
import CoreBluetooth
import FlutterMacOS

let GSS_SUFFIX = "0000-1000-8000-00805f9b34fb"

extension CBUUID {
    public var uuidStr: String {
        get {
            uuidString.lowercased()
        }
    }
}

extension CBPeripheral {
    public func getCharacteristic(_ characteristic: String, of service: String) -> CBCharacteristic {
        let s = self.services?.first {
            $0.uuid.uuidStr == service || "0000\($0.uuid.uuidStr)-\(GSS_SUFFIX)" == service
        }
        let c = s?.characteristics?.first {
            $0.uuid.uuidStr == characteristic || "0000\($0.uuid.uuidStr)-\(GSS_SUFFIX)" == characteristic
        }
        return c!
    }

    public func setNotifiable(_ bleInputProperty: String, for characteristic: String, of service: String) {
        setNotifyValue(bleInputProperty != "disabled", for: getCharacteristic(characteristic, of: service))
    }
}

public class NotepadCorePlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let plugin = NotepadCorePlugin()

        let channel = FlutterMethodChannel(name: "notepad_core/method", binaryMessenger: registrar.messenger)
        registrar.addMethodCallDelegate(plugin, channel: channel)

        FlutterEventChannel(name: "notepad_core/event.scanResult", binaryMessenger: registrar.messenger).setStreamHandler(plugin)
        plugin.connectorMessage = FlutterBasicMessageChannel(name: "notepad_core/message.connector", binaryMessenger: registrar.messenger)
        plugin.clientMessage = FlutterBasicMessageChannel(name: "notepad_core/message.client", binaryMessenger: registrar.messenger)
    }

    private var manager: CBCentralManager!
    private var discoveredPeripherals: Dictionary<String, CBPeripheral>!
    private var peripheral: CBPeripheral?

    private var serviceConfigGroup: DispatchGroup?

    private var scanResultSink: FlutterEventSink?
    private var connectorMessage: FlutterBasicMessageChannel!
    private var clientMessage: FlutterBasicMessageChannel!

    override init() {
        super.init()
        manager = CBCentralManager(delegate: self, queue: nil)
        discoveredPeripherals = Dictionary()
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("handleMethodCall \(call.method)")
        switch call.method {
        case "isBluetoothAvailable":
            result(manager.state == .poweredOn)
        case "startScan":
            manager.scanForPeripherals(withServices: nil)
            result(nil)
        case "stopScan":
            manager.stopScan()
            result(nil)
        case "connect":
            guard let arguments = call.arguments as? Dictionary<String, Any>,
                  let deviceId = arguments["deviceId"] as? String else {
                result(FlutterMethodNotImplemented)
                return
            }
            peripheral = discoveredPeripherals[deviceId]
            peripheral!.delegate = self
            manager.connect(peripheral!)
            result(nil)
        case "disconnect":
            if peripheral != nil {
                manager.cancelPeripheralConnection(peripheral!)
            }
            clean()
            result(nil)
        case "discoverServices":
            peripheral?.discoverServices(nil)
            result(nil)
        case "setNotifiable":
            guard let arguments = call.arguments as? Dictionary<String, Any>,
                  let service = arguments["service"] as? String,
                  let characteristic = arguments["characteristic"] as? String,
                  let bleInputProperty = arguments["bleInputProperty"] as? String else {
                result(FlutterMethodNotImplemented)
                return
            }
            peripheral?.setNotifiable(bleInputProperty, for: characteristic, of: service)
            result(nil)
        case "readValue":
            guard let arguments = call.arguments as? Dictionary<String, Any>,
                  let service = arguments["service"] as? String,
                  let characteristic = arguments["characteristic"] as? String else {
                result(FlutterMethodNotImplemented)
                return
            }
            peripheral?.readValue(for: peripheral!.getCharacteristic(characteristic, of: service))
            result(nil)
        case "writeValue":
            guard let arguments = call.arguments as? Dictionary<String, Any>,
                  let service = arguments["service"] as? String,
                  let characteristic = arguments["characteristic"] as? String,
                  let value = arguments["value"] as? FlutterStandardTypedData,
                  let bleOutputProperty = arguments["bleOutputProperty"] as? String else {
                result(FlutterMethodNotImplemented)
                return
            }
            let type = bleOutputProperty == "withoutResponse" ? CBCharacteristicWriteType.withoutResponse : CBCharacteristicWriteType.withResponse
            peripheral?.writeValue(value.data,
                    for: peripheral!.getCharacteristic(characteristic, of: service),
                    type: type)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

extension NotepadCorePlugin: CBCentralManagerDelegate {
    public func centralManagerDidUpdateState(_ central: CBCentralManager) {
        print("centralManagerDidUpdateState \(central.state.rawValue)")
        switch central.state {
        case .poweredOff:
            connectorMessage.sendMessage(["BluetoothState": "unavailable"])
        case .poweredOn:
            connectorMessage.sendMessage(["BluetoothState": "available"])
        default:
            break
        }
    }

    public func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber) {
        print("centralManager:didDiscoverPeripheral \(peripheral.name) \(peripheral.identifier)")
        discoveredPeripherals[peripheral.identifier.uuidString] = peripheral

        let manufacturerData = advertisementData[CBAdvertisementDataManufacturerDataKey] as? Data
        scanResultSink?([
            "name": peripheral.name ?? "",
            "deviceId": peripheral.identifier.uuidString,
            "manufacturerData": FlutterStandardTypedData(bytes: manufacturerData ?? Data()),
            "rssi": RSSI,
        ])
    }

    public func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        print("centralManager:didConnect \(peripheral.identifier)")
        if self.peripheral != peripheral {
            print("Probably MEMORY LEAK!")
            return
        }
        connectorMessage.sendMessage(["ConnectionState": "connected"])
    }

    public func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        print("centralManager:didDisconnectPeripheral: \(peripheral.identifier) error: \(error)")
        if self.peripheral != peripheral {
            print("Probably MEMORY LEAK!")
            return
        }
        clean()
        connectorMessage.sendMessage(["ConnectionState": "disconnected"])
    }

    private func clean() {
        peripheral = nil
    }
}

extension NotepadCorePlugin: FlutterStreamHandler {
    open func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        guard let args = arguments as? Dictionary<String, Any>, let name = args["name"] as? String else {
            return nil
        }
        print("NotepadCorePlugin onListenWithArguments：\(name)")
        if name == "scanResult" {
            scanResultSink = events
        }
        return nil
    }

    open func onCancel(withArguments arguments: Any?) -> FlutterError? {
        guard let args = arguments as? Dictionary<String, Any>, let name = args["name"] as? String else {
            return nil
        }
        print("NotepadCorePlugin onCancelWithArguments：\(name)")
        if name == "scanResult" {
            scanResultSink = nil
        }
        return nil
    }
}

extension NotepadCorePlugin: CBPeripheralDelegate {
    public func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        print("peripheral: \(peripheral.identifier) didDiscoverServices: \(error)")
        if self.peripheral != peripheral {
            print("Probably MEMORY LEAK!")
            return
        }
        serviceConfigGroup = DispatchGroup()
        for service in peripheral.services! {
            serviceConfigGroup?.enter()
            peripheral.discoverCharacteristics(nil, for: service)
        }
        serviceConfigGroup?.notify(queue: DispatchQueue.main) {
            self.serviceConfigGroup = nil
            self.connectorMessage.sendMessage(["ServiceState": "discovered"])
        }
    }

    public func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        for characteristic in service.characteristics! {
            print("peripheral:didDiscoverCharacteristicsForService (\(service.uuid), \(characteristic.uuid)")
        }
        if self.peripheral != peripheral {
            print("Probably MEMORY LEAK!")
            return
        }
        serviceConfigGroup?.leave()
    }

    public func peripheral(_ peripheral: CBPeripheral, didUpdateNotificationStateFor characteristic: CBCharacteristic, error: Error?) {
        print("peripheral:didUpdateNotificationStateFor \(characteristic.uuid) \(characteristic.isNotifying)")
        if self.peripheral != peripheral {
            print("Probably MEMORY LEAK!")
            return
        }
        clientMessage.sendMessage(["characteristicConfig": characteristic.uuid.uuidStr])
    }

    public func peripheral(_ peripheral: CBPeripheral, didWriteValueFor characteristic: CBCharacteristic, error: Error?) {
        let data = characteristic.value as? NSData
        print("peripheral:didWriteValueForCharacteristic \(characteristic.uuid) \(data) error: \(error)")
        if self.peripheral != peripheral {
            print("Probably MEMORY LEAK!")
            return
        }
    }

    public func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        let data = characteristic.value!
        print("peripheral:didUpdateValueForForCharacteristic \(characteristic.uuid) \(data as NSData) error: \(error)")
        if self.peripheral != peripheral {
            print("Probably MEMORY LEAK!")
            return
        }
        let characteristicValue: [String: Any] = [
            "characteristic": characteristic.uuid.uuidStr,
            "value": FlutterStandardTypedData(bytes: data)
        ]
        clientMessage.sendMessage(["characteristicValue": characteristicValue])
    }
}
